tools/clang-display-context/ClangDispCont.cpp

```diff
@@ -0,0 +1,169 @@
+    cl::cat(ClangDispContCategory));
+
+class DisplayContextConsumer : public ASTConsumer {
+public:
+  DisplayContextConsumer(ASTContext &Context) : Ctx(Context) {}
+
+  ~DisplayContextConsumer() {
+    // Flush results to standard output.
+    outs() << IssueString;
+  }
+
+  virtual void HandleTranslationUnit(ASTContext &Ctx) {
+    if (!Ctx.getDiagnostics().hasErrorOccurred())
+      handleDecl(Ctx.getTranslationUnitDecl());
+  }
+
+private:
+  template <typename T> 
+  bool checkIfInteresting(const T *N);
+  void handleDecl(const Decl *D);
+  void handleStmt(const Stmt *S);
+  void setEnclosingDecl(const Decl *D);
+  void setIssueString(const std::string &S);
+  
+  ASTContext &Ctx;
+  FullSourceLoc IssueLoc;
+  const Decl *EnclosingDecl;
+  std::string IssueString;
+};
+
+template <typename T>
+bool DisplayContextConsumer::checkIfInteresting(const T *Node) {
+  // Nodes are only interesting if they contain the line with the issue
+  if (!Node)
+    return false;
+
+  FullSourceLoc L1 = Ctx.getFullLoc(Node->getLocStart());
+  FullSourceLoc L2 = Ctx.getFullLoc(Node->getLocEnd());
+
+  return L1.isValid() && L2.isValid() &&
+         Line <= L2.getExpansionLineNumber() &&
+         L1.getExpansionLineNumber() <= Line;
+}
+
+void DisplayContextConsumer::handleDecl(const Decl *D) {
+  if (!D)
+    return;
+
+  if (const auto *TD = dyn_cast<TemplateDecl>(D))
+    D = TD->getTemplatedDecl();
+  
+  if (const auto *DC = dyn_cast<DeclContext>(D)) {
+    for (const Decl *D : DC->decls()) {
+      if (!checkIfInteresting<Decl>(D))
+        continue;
+      handleDecl(D);
+      return;
+    }
+  }
+  
+  setEnclosingDecl(D);
+  IssueLoc = Ctx.getFullLoc(D->getLocation()).getExpansionLoc();
+  
+  handleStmt(D->getBody());
+  setIssueString(clang::GetIssueString(Ctx.getSourceManager(), IssueLoc, "",
+                                       "", EnclosingDecl, Ctx.getLangOpts()));
+}
+
+void DisplayContextConsumer::handleStmt(const Stmt *S) {
+  if (!checkIfInteresting<Stmt>(S))
+    return;
+
+  IssueLoc = Ctx.getFullLoc(S->getLocStart()).getExpansionLoc();
+  
+  if (const LambdaExpr *LE = dyn_cast<LambdaExpr>(S))
+    EnclosingDecl = LE->getCallOperator();
+
+  for (const Stmt *Child : S->children())
+    handleStmt(Child);
+}
+
+void DisplayContextConsumer::setEnclosingDecl(const Decl *D) {
+  // If the issue is located in a DeclaratorDecl which is not
+  // a FunctionDecl, traverse up until the enclosing context is found.
+  // This function aims to solve issues especially with VarDecls.
+  while (dyn_cast<DeclaratorDecl>(D) && !dyn_cast<FunctionDecl>(D)) {
+    D = dyn_cast<Decl>(D->getDeclContext());
+  }
+  EnclosingDecl = D;
+}
+
+void DisplayContextConsumer::setIssueString(const std::string &SAIssue) {
+  // Since Location is provided as a range, the IssueString from Clang SA
+  // has to be modified by replacing the column number, and removing
+  // the first and last delimeters (CheckerName and BugType not provided)
```

Comment:
What are these delimiter characters? How possible it is for this regex to break later in development? If ClangSA output changes...?

Also `.` missing from end of comment.